Entity Delete GAWB-1666

jmthibault79 opened the pull request:
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

jmthibault79 committed:
enable entity delete endpoint and tests
- Hide entity and attributes
- Add deleted column to entity table and attributes
- Submission test
- reduce -> flatten
- add type annotations
- Add actual-delete test

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/DriverComponent.scala:
```diff
@@ -61,6 +62,10 @@ trait DriverComponent {
   }
+
+  def renameForHiding(name: String): String = {
+    name + "_" + DateTime.now().toString("yyyy-MM-dd_HH.mm.ss")
```
Can't use colons here because of `AttributeName`s which use colon delimiters

helgridly replied:
can you and @ansingh7115 sync to genericise "gimme the timestamped version of this thing"?

jmthibault79 replied:
Yes we agreed on this representation

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:
```diff
@@ -7,7 +7,6 @@ import org.broadinstitute.dsde.rawls.RawlsException
 import org.broadinstitute.dsde.rawls.model.WorkspaceAccessLevels.WorkspaceAccessLevel
 import org.broadinstitute.dsde.rawls.model._
 import org.joda.time.DateTime
-import slick.dbio.Effect.Read
```
I can revert this file change if you like, since I didn't make a functional change here

ansingh7115 replied:
Yes, keep it

jmthibault79 replied:
ok

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponentSpec.scala:
```diff
@@ -39,55 +39,55 @@ class AttributeComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers
     runAndWait(insertWorkspaceAttributeRecords(workspaceId, AttributeName.withDefaultNS("test"), testAttribute))
 
-    assertExpectedRecords(WorkspaceAttributeRecord(dummyId2, workspaceId, AttributeName.defaultNamespace, "test", Option("test"), None, None, None, None, None, None))
+    assertExpectedRecords(WorkspaceAttributeRecord(dummyId2, workspaceId, AttributeName.defaultNamespace, "test", Option("test"), None, None, None, None, None, None, false))
```
This file has only changes like this

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala:
```diff
@@ -70,19 +70,9 @@ class EntityComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers wit
-    assertResult(false) { runAndWait(entityQuery.delete(workspaceContext, "type", "name")) }
-  }
-
-  it should "get entity types" in withDefaultTestDatabase {
```
`getEntityTypes()` was not used, so I removed it

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala:
```diff
@@ -92,13 +82,25 @@ class EntityComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers wit
+      }
+      runAndWait(deleteSamples)
 
+      assertResult(Map("PairSet" -> 1, "Individual" -> 2, "Aliquot" -> 2, "SampleSet" -> 5, "Pair" -> 2)) {
```
Same as line 86, minus Samples

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/TestDriverComponent.scala:
```diff
@@ -713,7 +713,6 @@ trait TestDriverComponent extends DriverComponent with DataAccess {
     val sample6 = Entity("sample6", "Sample", Map(AttributeName.withDefaultNS("type") -> AttributeString("tumor")))
     val sample7 = Entity("sample7", "Sample", Map(AttributeName.withDefaultNS("type") -> AttributeString("tumor"), AttributeName.withDefaultNS("cycle") -> AttributeEntityReference("Sample", "sample6")))
     val sample8 = Entity("sample8", "Sample", Map(AttributeName.withDefaultNS("type") -> AttributeString("tumor")))
-    val extraSample = Entity("extraSample", "Sample", Map.empty)
```
not used anywhere

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala:
```diff
@@ -278,6 +278,34 @@ class SubmissionSpec(_system: ActorSystem) extends TestKit(_system) with FlatSpe
     checkSubmissionStatus(workspaceService, newSubmissionReport.submissionId)
   }
 
+  it should "continue to monitor a Submission on a deleted entity" in withWorkspaceServiceMockExecution { mockExecSvc => workspaceService =>
```
Submissions refer to Entities by ID, which will continue to reference the same DB record

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiServiceSpec.scala:
```diff
@@ -723,23 +1319,29 @@ class EntityApiServiceSpec extends ApiServiceSpec {
-  /*
-    following 2 tests disabled until entity deletion is re-implemented (see GAWB-422)
-   */
-  ignore should "*DISABLED* return 204 on entity delete" in withTestDataApiServices { services =>
```
These basic deletion tests have been moved upward in this file.  The green "changes" are new tests.

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:
```diff
@@ -569,14 +569,6 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
           status
         }
       }
-    //this endpoint is only temporarily returning MethodNotAllowed. When GAWB-422 is complete, it should return NotFound
```
This isn't a necessary part of this test

jmthibault79 commented:
Discussed FISMA with DevOps and Comms - will have to make a note in the user docs

jmthibault79 commented:
TODO:
~- update AC in JIRA~
~- test in FIAB~
- Orch Swagger and Endpoint

ansingh7115 commented on core/src/main/resources/swagger/rawls.yaml:
```diff
@@ -2189,50 +2189,58 @@ paths:
-#            - email
-#            - profile
+  '/api/workspaces/{workspaceNamespace}/{workspaceName}/entities/delete':
+    post:
```
Why is this a post and not a delete?

jmthibault79 replied:
So it can take a JSON payload

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala:
```diff
@@ -220,12 +227,12 @@ trait AttributeComponent {
 
     def marshalAttributeEmptyEntityReferenceList(ownerId: OWNER_ID, attributeName: AttributeName): RECORD = {
-      createRecord(0, ownerId, attributeName.namespace, attributeName.name, None, None, None, None, None, None, Option(0))
+      createRecord(0, ownerId, attributeName.namespace, attributeName.name, None, None, None, None, None, None, Option(0), false)
```
are these ever used to update a record that's already been deleted?

jmthibault79 replied:
Good point.  I will need to investigate that some more.

jmthibault79 replied:
Good news everyone: they are not

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:
```diff
@@ -184,21 +189,23 @@ trait EntityComponent {
 
     def findEntityByWorkspace(workspaceId: UUID): EntityQuery = {
-      filter(_.workspaceId === workspaceId)
+      filter(entRec => entRec.workspaceId === workspaceId && ! entRec.deleted)
```
Maybe rename findEntityByType to findActiveEntityByType or something like that? (same for findEntityByWorkspace and getEntityTypesWithCounts?)

helgridly replied:
yeah, i was wondering this as well. some way to distinguish between the fns that do and don't retried deleteds

jmthibault79 replied:
I agree that this is necessary, but I'd prefer to do that as part of a bigger EntityComponent refactoring: https://broadinstitute.atlassian.net/browse/GAWB-1768

helgridly replied:
If you advocate to get said ticket prio'd soon then I'm fine with that.

jmthibault79 replied:
Can do

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:
```diff
@@ -216,7 +224,7 @@ trait EntityComponent {
     }
 
-    /** gets the given entity */
+    // includes "deleted" hidden entities when using their hidden names
```
Yay for all these comments!

ansingh7115 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponent.scala:
```diff
@@ -173,7 +172,7 @@ trait MethodConfigurationComponent {
     def hideMethodConfigurationAction(id: Long, methodConfigName: String): ReadWriteAction[Int] = {
       findById(id).map(rec => (rec.deleted, rec.name))
-        .update(true, methodConfigName + "_" + DateTime.now().toString("yyyy-MM-dd_HH:mm:ss"))
+        .update(true, renameForHiding(methodConfigName))
```
Excellent!

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:
```diff
@@ -468,32 +477,46 @@ trait EntityComponent {
-
-      referencesResults.map(_.reduce(_ ++ _)).flatMap { refIdOptions =>
-        val refIds = refIdOptions.collect { case Some(id) => id }.toSet
+      DBIO.sequence(batchActions).map(_.reduce(_ ++ _)).flatMap { refIds =>
```
this... reduces all the results from the batches into a single list, rather than a list-of-lists?

jmthibault79 replied:
Yes... which is the same as flatten, so I'll change it to flatten for clarity.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -881,11 +881,18 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
-        withEntity(workspaceContext, entityType, entityName, dataAccess) { entity =>
-          dataAccess.entityQuery.delete(workspaceContext, entity.entityType, entity.name).map(_ => RequestComplete(StatusCodes.NoContent))
+        withAllEntities(workspaceContext, dataAccess, entRefs) { entities =>
+          dataAccess.entityQuery.getAllReferringEntities(workspaceContext, entRefs.toSet) flatMap { referringEntities =>
```
can all of this move to a *Component somewhere? it feels a bit too databasey for a *Service

jmthibault79 replied:
counterpoint: Components don't know about PerRequest.  This doesn't seem out of line with the rest of WorkspaceService to me

jmthibault79 replied:
The "refactor EntityComponent.hide()" commit makes this less databasey.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1858,13 +1865,40 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+
+    DBIO.sequence(entityActions) flatMap { entityTries =>
+      val failures = entityTries.collect { case Failure(y) => y.getMessage }
+      if (failures.size == entityTries.size) {
```
i think you can reasonably merge this block with the one below, and return the full list in the case where none of them can be found

jmthibault79 replied:
This division was suggested by Alex.  I'm good with either way.  We can chat.

jmthibault79 replied:
@helgridly we're OK with collapsing these two cases.  400 or 404?

helgridly replied:
@breilly2 has convinced me that 400 is the right answer, because 404 implies the resource defined by the URL has not been found.

helgridly commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala:
```diff
@@ -70,19 +70,9 @@ class EntityComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers wit
-        }
-      }
-
+    assertResult(false) { runAndWait(entityQuery.hide(workspaceContext, "type", "name")) }
```
now this is switched to hide, we need tests to make sure that actual-delete actually deletes

helgridly commented on core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiServiceSpec.scala:
```diff
@@ -540,6 +1037,76 @@ class EntityApiServiceSpec extends ApiServiceSpec {
       }
   }
 
+  it should "return 200 on get deleted entity accessed by its hidden name" in withTestDataApiServices { services =>
```
what's this used for in practise? historical submissions?

jmthibault79 replied:
That was my thought: enable click-navigation of deleted Entities via their Submissions

helgridly reviewed (changes requested):
this is really impressive! mostly a few nitpicks / questions

jmthibault79 commented:
~Hold on... my FIAB testing shows this to be quite broken.~
Fixed it

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:
```diff
@@ -184,21 +189,23 @@ trait EntityComponent {
 
     def entityAttributes(entityId: Rep[Long]): EntityAttributeQuery = for {
-      entityAttrRec <- entityAttributeQuery if entityAttrRec.ownerId === entityId
+      entityAttrRec <- entityAttributeQuery if entityAttrRec.ownerId === entityId && ! entityAttrRec.deleted
```
this looks fishy. if thing == other thing and [bool value]?

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:
```diff
@@ -184,21 +189,23 @@ trait EntityComponent {
 
     def entityAttributes(entityId: Rep[Long]): EntityAttributeQuery = for {
-      entityAttrRec <- entityAttributeQuery if entityAttrRec.ownerId === entityId
+      entityAttrRec <- entityAttributeQuery if entityAttrRec.ownerId === entityId && ! entityAttrRec.deleted
```
is this your bug? this is either hard to read or it's a bit funny. shouldn't there be a filter in here? `entityAttributeQuery.filter( _.entity_id == entityId )` or something? where is `entityAttrRec` coming from on the RHS here?

jmthibault79 replied:
`entityAttrRec` works like a function parameter here, so this should be equivalent to `entityAttributeQuery.filter(_.owner_id === entityId).filterNot(_.deleted)`

jmthibault79 replied:
This was not the bug.

helgridly replied:
TIL. I find this syntax weird but I'll deal with it.

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:
```diff
@@ -338,18 +353,32 @@ trait EntityComponent {
+    }
+
+    private def hideEntityAttribute(attrRec: EntityAttributeRecord): WriteAction[Int] = {
+      entityAttributeQuery.filter(_.id === attrRec.id).map(rec => (rec.deleted, rec.name)).update(true, renameForHiding(attrRec.name))
```
The bug was here: I was filtering on attribute name.  So hiding an entity with a "normal" attribute would hide ALL entities' "normal" attributes.  Haha, oops.

helgridly replied:
I assume the tests you added to discover this do a good job of checking for such issues in the future?

jmthibault79 replied:
The last test in EntityComponentSpec (the one with Bob) reproduces the problem, and the added test calls to `countEntitiesAttrs` are intended to confirm that the state of the world is sane.

helgridly replied:
Great, thanks.

jmthibault79 commented:
Orchestration PR at https://github.com/broadinstitute/firecloud-orchestration/pull/403

Addressed comments except for https://github.com/broadinstitute/rawls/pull/644/files#r105736254 and https://github.com/broadinstitute/rawls/pull/644/files#r105758751.  Do you want to discuss these further?  Ready for re-review.

jmthibault79 committed:
add millis to renameForHiding

helgridly reviewed (approved):
👍 pending 404 collapse. Go ahead and merge once that's done and you're happy.

helgridly commented:
(though Travis is currently compile-erroring)
